src/System.Private.CoreLib/src/System/Decimal.DecCalc.cs

```diff
@@ -2239,52 +2231,60 @@ internal static unsafe void VarDecDiv(ref Decimal d1, ref Decimal d2)
             //**********************************************************************
             // VarDecMod - Computes the remainder between two decimals
             //**********************************************************************
-            internal static Decimal VarDecMod(ref Decimal d1, ref Decimal d2)
+            internal static void VarDecMod(ref Decimal d1, ref Decimal d2)
             {
-                // OleAut doesn't provide a VarDecMod.            
+                if ((d2.lo | d2.mid | d2.hi) == 0)
+                    throw new DivideByZeroException();
+
+                if ((d1.lo | d1.mid | d1.hi) == 0)
+                    return;
 
                 // In the operation x % y the sign of y does not matter. Result will have the sign of x.
                 d2.uflags = (d2.uflags & ~SignMask) | (d1.uflags & SignMask);
 
+                int cmp = VarDecCmpSub(ref d1, ref d2);
+                if (cmp == 0)
+                {
+                    d1.lo = 0;
+                    d1.mid = 0;
+                    d1.hi = 0;
+                    if (d2.uflags > d1.uflags)
+                        d1.uflags = d2.uflags;
+                    return;
+                }
+                if ((cmp ^ (int)(d1.uflags & SignMask)) < 0)
+                    return;
 
                 // This piece of code is to work around the fact that Dividing a decimal with 28 digits number by decimal which causes
                 // causes the result to be 28 digits, can cause to be incorrectly rounded up.
                 // eg. Decimal.MaxValue / 2 * Decimal.MaxValue will overflow since the division by 2 was rounded instead of being truncked.
-                if (Math.Abs(d1) < Math.Abs(d2))
-                {
-                    return d1;
-                }
-                d1 -= d2;
-
-                if (d1 == 0)
-                {
-                    // The sign of D1 will be wrong here. Fall through so that we still get a DivideByZeroException
-                    d1.uflags = (d1.uflags & ~SignMask) | (d2.uflags & SignMask);
-                }
+                Decimal tmp = d2;
+                DecAddSub(ref d1, ref tmp, true);
```

Comment:
Nit: `DecAddSub` trashing the second argument is non-intuitive ... it may be useful to add comment about it in front of `DecAddSub` at least.

Reply:
And in the readonly decimal, use `ref` for arguments that can be trashed; and `in` for arguments that  stay intact?

Reply:
I'll add comments for parameter trashing and use `in` where possible (can't use `in` for mutable structs because property calls would require implicit local copies).